tedge/src/cli/mqtt/subscribe.rs

```diff
@@ -26,57 +24,52 @@ impl Command for MqttSubscribeCommand {
     }
 }
 
-#[tokio::main]
-async fn subscribe(cmd: &MqttSubscribeCommand) -> Result<(), MqttError> {
-    let config = cmd.mqtt_config.clone().clean_session();
-    let mqtt = Client::connect(cmd.client_id.as_str(), &config).await?;
-    let filter = TopicFilter::new(cmd.topic.as_str())?.qos(cmd.qos);
+fn subscribe(cmd: &MqttSubscribeCommand) -> Result<(), MqttError> {
+    let mut options = MqttOptions::new(
+        cmd.client_id.as_str(),
+        &cmd.mqtt_config.host,
+        cmd.mqtt_config.port,
+    );
+    options.set_clean_session(true);
 
-    let mut errors = mqtt.subscribe_errors();
-    let mut messages = mqtt.subscribe(filter).await?;
+    let (mut client, mut connection) = Client::new(options, cmd.mqtt_config.queue_capacity);
 
-    loop {
-        select! {
-            error = errors.next().fuse() => {
-                if let Some(err) = error {
-                    if err.to_string().contains("MQTT connection error: I/O: Connection refused (os error 111)") {
-                        return Err(MqttError::ServerError(err.to_string()));
+    for event in connection.iter() {
```

Comment:
Where will the reconnect on a disconnect happen? This loop won't cover that, would it?

Reply:
That's the magic part compared to https://github.com/thin-edge/thin-edge.io/pull/417.

On TCP errors, the rumqtt library automatically reconnect on the background. This is the case here and in https://github.com/thin-edge/thin-edge.io/pull/417. __But__, the new state is incomplete when the session is clean: one has no more subscription. Here, the fix is simple. One just have to re-subscribe when the connection is acknowledged at the MQTT level (with a `ConnAck`).  In https://github.com/thin-edge/thin-edge.io/pull/417 we have no way to react on `ConnAck`, hence a disconnect is forced to restart with a fresh new MQTT client.